### ISSUE TITLE
Need an available way to finish formatting works on Windows

### ISSUE THREAD

**Issue by TheBlueAlgae** [google/comprehensive-rust#1394]
Recently I attempted updating translation files (also the `.po` file in `/po` folder) and ran `dprint fmt` on my computer, but when I committed the file, I surprisingly found that the file did not format as the proper form that the repository showed.

Later, @mgeisler told me that the way the `msgcat` handles translation files is different between Windows and Linux. He also told me to try some values that matches the standard format. Ok, well, I not only tried 79 and other adjacent numbers, but also ran `msgcat -o po/xx.po --no-wrap po/xx.po`.

But unfortunately, none of these methods have worked.

So, I determined to open this issue after attempting varieties of methods. I hope someone can find out some ways to solve this issue, or it might make contributors who use Windows confused.

**Comment by mgeisler** [comment]
Hi @TheBlueAlgae,

Thanks for opening this! We should track the findings of #1381 here so we can find a solution.

> Later, @mgeisler told me that the way the `msgcat` handles translation files is different between Windows and Linux. He also told me to try some values that matches the standard format. Ok, well, I not only tried 79 and other adjacent numbers, but also ran `msgcat -o po/xx.po --no-wrap po/xx.po`.

More precisely: I'm surprised to see that `msgcat` apparently works different between the two platforms. That was not what I hoped for since it complicates things a lot for contributors like yourself!

## Status Right Now

I'm on Linux and use `msgcat` from the [gettext-0.21-13 package in Debian](https://packages.debian.org/trixie/gettext). We do the same when we test translations, we get Gettext version 0.21 from Ubuntu.

This seems to work well: `dprint check` reports that all files are formatted correctly.

I will try installing Gettext on a Windows machine somehow. That will allow me to debug what is going on.

In parallel, you could try using [Windows Subsystem for Linux (WSL)](https://learn.microsoft.com/en-us/windows/wsl/install) to install Gettext and the rest of the tools. The default Linux distrubtion is Ubuntu, so it should work very well out of the box.

**Comment by mgeisler** [comment]
> I will try installing Gettext on a Windows machine somehow. That will allow me to debug what is going on.

I downloaded the [Gettext binaries for Windows](https://mlocati.github.io/articles/gettext-iconv-windows.html), I picked the [gettext0.21-iconv1.16-static-64.zip](https://github.com/mlocati/gettext-iconv-windows/releases/download/v0.21-v1.16/gettext0.21-iconv1.16-static-64.zip) to try on a laptop with a recent Windows.

I then downloaded `da.po` from this repository and I can reproduce the problem: There are subtle differences in how the file is formatted after running it through `msgcat`! There were 30 differences in the Danish file I used — they were differences in the wrapping of the English `msgid` fields. It looks like a change in how special characters such as `\`, `+`, etc are handled.

Next, I tried downloading a different set of Gettext tools: [GNU gettext tools for Windows](https://github.com/vslavik/gettext-tools-windows/releases). The `msgcat` binary from the latest 0.22 release gives me the same result as what you find in the repository! The binaries there are maintained by the same author who wrote Poedit, so I'm sure they'll be maintained well.

Can you try installing the binaries from that repository instead and tell me if `dprint fmt` works for you then? If so, we can update the contribution instructions to recommend that repository instead.

**Comment by TheBlueAlgae** [comment]
> Can you try installing the binaries from that repository instead and tell me if `dprint fmt` works for you then? If so, we can update the contribution instructions to recommend that repository instead.

To avoid breaking the existing unmerged commits, I will try installing this after my commits being merged.

**Comment by TheBlueAlgae** [comment]
> Can you try installing the binaries from that repository instead and tell me if `dprint fmt` works for you then? If so, we can update the contribution instructions to recommend that repository instead.

Hi @mgeisler !

I feel so happy that the second version of [Gettext Tool](https://github.com/vslavik/gettext-tools-windows/releases) you mentioned seems worked well! I used it to format an edited translation file and then made a commit. The file is now passed all checks and is ready to be merged into main branch!

However, I have not verified whether the `dprint` on Windows can also work well. I directly used `msgcat -o zh-CN.po zh-CN.po` to test.

In addition, this Gettext Tool does not have a setup program, with some compressions distributed instead. So, after downloading a compression distribution from the Release download page, contributors must decompress it into a directory before using this tool. Then, it is better that add the path where the `bin` folder is to system environment variables.

**Comment by mgeisler** [comment]
> I feel so happy that the second version of [Gettext Tool](https://github.com/vslavik/gettext-tools-windows/releases) you mentioned seems worked well! I used it to format an edited translation file and then made a commit. The file is now passed all checks and is ready to be merged into main branch!

Amazing, this is great news! Thanks a lot for verifying this :smile: 



> In addition, this Gettext Tool does not have a setup program, with some compressions distributed instead. So, after downloading a compression distribution from the Release download page, contributors must decompress it into a directory before using this tool. Then, it is better that add the path where the `bin` folder is to system environment variables.

It's been a few years since I used Windows regularly... but what you say make sense. Do you think you could write up instructions for where to put the files (if there is a directly commonly found in `%PATH%`?) or how to update the `%PATH%` if that's easier?

We now have both [Formatting](https://github.com/google/comprehensive-rust/blob/main/CONTRIBUTING.md#formatting) and [Preparation](https://github.com/google/comprehensive-rust/blob/main/TRANSLATIONS.md#preparation) as places where we explain how to install the necessary tooling. I would suggest that we unify this into a single location (I'm thinking `CONTRIBUTING.md`) and then point people to that file from both `TRANSLATIONS.md` and `README.md` etc.

**Comment by mgeisler** [comment]
I created #1426 to remind ourselves to simplify the setup instructions into one place.
